std/algorithm/iteration.d

```diff
@@ -4280,3 +4284,218 @@ unittest
          [1, 2, 0],
          [2, 1, 0]]));
 }
+
+/**
+Finds the mean (colloquially known as the average) of a range. This function
+will return $(D real.nan) if the range is empty. This function is
+$(BIGOH r.length).
+
+Params:
+    r = an input range
+
+Returns:
+    the mean of r
+*/
+real mean(R)(R r)
+if (isInputRange!R &&
+    !isInfinite!R &&
+    is(ElementType!R : real))
+{
+    if (r.empty)
+    {
+        return real.nan;
+    }
+
+    static if (hasLength!R)
+    {
+        if (r.length == 1)
+        {
+            return r.front;
+        }
+
+        return sum(r) / real(r.length);
+    }
+    else
+    {
+        import std.typecons : tuple;
+
+        auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))
```

Comment:
It would be nice if we could have `sum`'s accurate summation for the non-length case too - it does support it - but I understand why that's not practical right now. Just wanted to point out that this path is less accurate.